types/src/account_state/mod.rs

```diff
@@ -21,6 +27,31 @@ impl AccountState {
             .map_err(Into::into)
     }
 
+    pub fn get_discovery_set_resource(&self) -> Result<Option<DiscoverySetResource>> {
+        self.0
+            .get(&*DISCOVERY_SET_RESOURCE_PATH)
+            .map(|bytes| lcs::from_bytes(bytes))
+            .transpose()
+            .map_err(Into::into)
+    }
+
+    pub fn get_event_handle_by_query_path(&self, query_path: &[u8]) -> Result<Option<EventHandle>> {
+        let event_handle = if *ACCOUNT_RECEIVED_EVENT_PATH == query_path {
+            self.get_account_resource()?
+                .map(|account_resource| account_resource.received_events().clone())
+        } else if *ACCOUNT_SENT_EVENT_PATH == query_path {
+            self.get_account_resource()?
+                .map(|account_resource| account_resource.sent_events().clone())
+        } else if *DISCOVERY_SET_CHANGE_EVENT_PATH == query_path {
+            self.get_discovery_set_resource()?
+                .map(|discovery_set_resource| discovery_set_resource.change_events().clone())
+        } else {
+            bail!("Unrecognized query path: {:?}", query_path);
+        };
```

Comment:
why not use `match`?

Reply:
I wasn't able to get the compiler happy with my `match`s... ( you can't even put  `query_path` on the left side of the `==` operator)

Reply:
This sounds weird to me...

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3ac7648a6ec1a4a59dd4b777d460eb6f

Looks like you can't put a variable name as a pattern on match legs, even if it's a static ^